libgadget/utils/system.c

```diff
@@ -458,66 +458,41 @@ int MPI_Alltoallv_sparse(void *sendbuf, int *sendcnts, int *sdispls,
     return 0;
 }
 
+/* return the number of hosts */
 int
-cluster_get_hostid()
+cluster_get_num_hosts(void)
 {
     /* Find a unique hostid for the computing rank. */
-    char hostname[1024];
-    int i;
-    gethostname(hostname, 1024);
-
-    MPI_Barrier(MPI_COMM_WORLD);
-
-    int l = strlen(hostname) + 4;
-    int ml = 0;
     int NTask;
     int ThisTask;
-    char * buffer;
-    int * nid;
     MPI_Comm_size(MPI_COMM_WORLD, &NTask);
     MPI_Comm_rank(MPI_COMM_WORLD, &ThisTask);
-    MPI_Allreduce(&l, &ml, 1, MPI_INT, MPI_MAX, MPI_COMM_WORLD);
-
-    /* Avoid ta_malloc since this can be called very early. */
-    buffer = malloc(ml * NTask);
-    nid = malloc(sizeof(int) * NTask);
-    MPI_Allgather(hostname, ml, MPI_BYTE, buffer, ml, MPI_BYTE, MPI_COMM_WORLD);
-
-    typedef int(*compar_fn)(const void *, const void *);
-    qsort(buffer, NTask, ml, (compar_fn) strcmp);
-
-    nid[0] = 0;
-    for(i = 1; i < NTask; i ++) {
-        if(strcmp(buffer + i * ml, buffer + (i - 1) *ml)) {
-            nid[i] = nid[i - 1] + 1;
-        } else {
-            nid[i] = nid[i - 1];
-        }
-    }
-    for(i = 0; i < NTask; i ++) {
-        if(!strcmp(hostname, buffer + i * ml)) {
-            break;
+
+    /* Size is set by the minimal size of the temp heap*/
+    const int bufsz = 256;
+    char * buffer = ta_malloc("buffer", char, bufsz * NTask);
```

Comment:
This almost overflows ta_malloc's limit. Perhaps add a comment about the crossref between bufsz and ta_malloc_init.